File: HisabCore/Sources/HisabCore/SyntheticCSVParser.swift
import Foundation

/// Demo/test statement format. Header line "hisab-demo-csv,v1"; optional
/// "period,yyyy-MM-dd,yyyy-MM-dd" second line; rows:
/// date,amountPaise,debit|credit,counterparty,reference,narration
public struct SyntheticCSVParser: StatementParser {
    public let source = Source.gpay

    public init() {}

    private static let dateFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.calendar = YearMonth.istCalendar
        formatter.timeZone = YearMonth.istCalendar.timeZone
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter
    }()

    public func canParse(data: Data, filename: String) -> Bool {
        guard let text = String(data: data, encoding: .utf8) else { return false }
        return text.hasPrefix("hisab-demo-csv,v1")
    }

    public func parse(data: Data, password: String?) throws -> ParsedDocument {
        guard let text = String(data: data, encoding: .utf8), text.hasPrefix("hisab-demo-csv,v1") else {
            throw ParseError.unrecognizedFormat
        }
        var lines = text.split(separator: "\n", omittingEmptySubsequences: true)
            .map { $0.trimmingCharacters(in: .whitespaces) }
        lines.removeFirst()  // header

        var declaredPeriod: DatePeriod?
        if let first = lines.first, first.hasPrefix("period,") {
            let parts = first.split(separator: ",", omittingEmptySubsequences: false)
            guard parts.count == 3,
                  let start = Self.dateFormatter.date(from: String(parts[1])),
                  let end = Self.dateFormatter.date(from: String(parts[2])) else {
                throw ParseError.malformedRow(2, first)
            }
            declaredPeriod = DatePeriod(start: start, end: end)
            lines.removeFirst()
        }

        guard !lines.isEmpty else { throw ParseError.empty }

        var transactions: [ParsedTransaction] = []
        for (offset, line) in lines.enumerated() {
            let fields = line.split(separator: ",", omittingEmptySubsequences: false).map(String.init)
            guard fields.count == 6,
                  let date = Self.dateFormatter.date(from: fields[0]),
                  let amount = Int64(fields[1]), amount > 0,
                  let direction = Direction(rawValue: fields[2]) else {
                throw ParseError.malformedRow(offset + (declaredPeriod == nil ? 2 : 3), line)
            }
            transactions.append(ParsedTransaction(
                date: date, amountPaise: amount, direction: direction,
                counterparty: fields[3],
                reference: fields[4].isEmpty ? nil : fields[4],
                narration: fields[5]))
        }
        return ParsedDocument(source: source, declaredPeriod: declaredPeriod, transactions: transactions)
    }
}

File: hisab_flutter/packages/hisab_core/lib/src/parsers/synthetic_csv_parser.dart
/// Demo/test statement format. Port of SyntheticCSVParser.swift.
/// Header "hisab-demo-csv,v1"; optional "period,yyyy-MM-dd,yyyy-MM-dd";
/// rows: date,amountPaise,debit|credit,counterparty,reference,narration
library;

import 'dart:convert';

import '../domain.dart';
import '../generic_bank/statement_date.dart';

class SyntheticCsvParser implements StatementParser {
  @override
  final Source source;

  const SyntheticCsvParser({this.source = Source.gpay});

  static String? _decode(List<int> data) {
    try {
      return utf8.decode(data);
    } on FormatException {
      return null;
    }
  }

  @override
  bool canParse(List<int> data, String filename) =>
      _decode(data)?.startsWith('hisab-demo-csv,v1') ?? false;

  @override
  ParsedDocument parse(List<int> data, {String? password}) {
    final text = _decode(data);
    if (text == null || !text.startsWith('hisab-demo-csv,v1')) {
      throw const UnrecognizedFormatException();
    }
    final lines = text
        .split('\n')
        .map((l) => l.trim())
        .where((l) => l.isNotEmpty)
        .toList()
      ..removeAt(0);

    DatePeriod? declaredPeriod;
    if (lines.isNotEmpty && lines.first.startsWith('period,')) {
      final parts = lines.first.split(',');
      final start =
          parts.length == 3 ? StatementDate.tryParse(parts[1], 'yyyy-MM-dd') : null;
      final end =
          parts.length == 3 ? StatementDate.tryParse(parts[2], 'yyyy-MM-dd') : null;
      if (start == null || end == null) {
        throw MalformedRowException(2, lines.first);
      }
      declaredPeriod = DatePeriod(start, end);
      lines.removeAt(0);
    }
    if (lines.isEmpty) throw const EmptyDocumentException();

    final transactions = <ParsedTransaction>[];
    for (var offset = 0; offset < lines.length; offset++) {
      final line = lines[offset];
      final fields = line.split(',');
      final date = fields.isNotEmpty
          ? StatementDate.tryParse(fields[0], 'yyyy-MM-dd')
          : null;
      final amount = fields.length > 1 ? int.tryParse(fields[1]) : null;
      final direction = fields.length > 2
          ? {'debit': Direction.debit, 'credit': Direction.credit}[fields[2]]
          : null;
      if (fields.length != 6 ||
          date == null ||
          amount == null ||
          amount <= 0 ||
          direction == null) {
        throw MalformedRowException(
            offset + (declaredPeriod == null ? 2 : 3), line);
      }
      transactions.add(ParsedTransaction(
        date: date,
        amountPaise: amount,
        direction: direction,
        counterparty: fields[3],
        reference: fields[4].isEmpty ? null : fields[4],
        narration: fields[5],
      ));
    }
    return ParsedDocument(
        source: source,
        declaredPeriod: declaredPeriod,
        transactions: transactions);
  }
}
